src/sqlancer/hsqldb/gen/HSQLDBUpdateGenerator.java

```diff
@@ -4,54 +4,53 @@
 import sqlancer.common.query.SQLQueryAdapter;
 import sqlancer.hsqldb.HSQLDBErrors;
 import sqlancer.hsqldb.HSQLDBProvider;
 import sqlancer.hsqldb.HSQLDBSchema;
+import sqlancer.hsqldb.HSQLDBSchema.HSQLDBColumn;
 import sqlancer.hsqldb.HSQLDBToStringVisitor;
 import sqlancer.hsqldb.ast.HSQLDBExpression;
 
-public final class HSQLDBUpdateGenerator {
+public final class HSQLDBUpdateGenerator extends AbstractUpdateGenerator<HSQLDBColumn> {
 
-    private static final ExpectedErrors EXPECTED_ERRORS = new ExpectedErrors();
     private final HSQLDBProvider.HSQLDBGlobalState globalState;
+    private final ExpectedErrors errors = new ExpectedErrors();
+    private HSQLDBExpressionGenerator gen;
 
     private HSQLDBUpdateGenerator(HSQLDBProvider.HSQLDBGlobalState globalState) {
         this.globalState = globalState;
     }
 
     public static SQLQueryAdapter getQuery(HSQLDBProvider.HSQLDBGlobalState globalState) {
-        return new HSQLDBUpdateGenerator(globalState).getQuery();
+        return new HSQLDBUpdateGenerator(globalState).generate();
     }
 
-    private SQLQueryAdapter getQuery() {
-        StringBuilder sb = new StringBuilder("UPDATE ");
+    private SQLQueryAdapter generate() {
         HSQLDBSchema.HSQLDBTable table = globalState.getSchema().getRandomTable(t -> !t.isView());
+        List<HSQLDBSchema.HSQLDBColumn> columns = table.getRandomNonEmptyColumnSubset();
+        gen = new HSQLDBExpressionGenerator(globalState).setColumns(table.getColumns());
+        sb.append("UPDATE ");
         sb.append(table.getName());
-        HSQLDBExpressionGenerator gen = new HSQLDBExpressionGenerator(globalState).setColumns(table.getColumns());
         sb.append(" SET ");
-        ExpectedErrors errors = new ExpectedErrors();
-        List<HSQLDBSchema.HSQLDBColumn> columns = table.getRandomNonEmptyColumnSubset();
-        for (int i = 0; i < columns.size(); i++) {
-            if (i != 0) {
-                sb.append(", ");
-            }
-            sb.append(columns.get(i).getName());
-            sb.append("=");
-            Node<HSQLDBExpression> expr;
-            expr = gen.generateConstant(columns.get(i).getType());
-            sb.append(HSQLDBToStringVisitor.asString(expr));
-            if (Randomly.getBooleanWithSmallProbability()) {
-                sb.append(" WHERE ");
-                expr = gen.generateExpression(columns.get(i).getType());
-                sb.append(HSQLDBToStringVisitor.asString(expr));
-                errors.add("data type of expression is not boolean");
-                HSQLDBErrors.addExpressionErrors(errors);
-            }
+        updateColumns(columns);
+        return new SQLQueryAdapter(sb.toString(), errors);
+    }
 
+    @Override
+    protected void updateValue(HSQLDBColumn column) {
+        Node<HSQLDBExpression> expr;
+        expr = gen.generateConstant(column.getType());
+        sb.append(HSQLDBToStringVisitor.asString(expr));
+        if (Randomly.getBooleanWithSmallProbability()) {
```

Comment:
I think this actually goes after the `updateColumn` call (as an `UPDATE` statement has only a single `WHERE` clause at the end of the statement). This seems to be a bug in the code before your refactoring (cc @nish-d). Perhaps you can fix this as part of this refactoring?